src/codecs/mod.rs

```diff
@@ -0,0 +1,18 @@
+#[cfg(test)]
+mod noop;
+
+use crate::config::DataType;
+#[cfg(test)]
+pub use noop::NoopCodec;
+use vector_core::{event::Event, transform::Transform};
+
+#[typetag::serde(tag = "type")]
+pub trait Codec: std::fmt::Debug + Send + Sync {
+    fn name(&self) -> &'static str;
+
+    fn build_decoder(&self) -> crate::Result<(Transform<Event>, DataType, DataType)>;
+
+    fn build_encoder(&self) -> crate::Result<(Transform<Event>, DataType, DataType)>;
+}
```

Comment:
Does it make sense to combine these two? That is, will sources ever use encoders, or sinks use decoders? Will we ever have an encoder for which a decoder is either not used in vector or not easily implemented, or vice versa? This is certainly more convenient for maintaining an inventory where there are both.

Reply:
Good question, I asked myself the same thing! 😄

> Will we ever have an encoder for which a decoder is either not used in vector or not easily implemented, or vice versa?

I can't come up with any, and if there was a codec where only one side makes sense, the other one could return an `Err` with "not implemented" (that should never be reached).

Reply:
Hm, I'm also leaning to combining them now. The `build` function could take a hint if it should decode or encode, based on if it is used from a source or sink. However, this hint could be overridden if the codec implements an explicit option to specify the decoding/encoding side.

Reply:
Oh, my comment about combining them was not to suggest we have only one build function, but rather about combining them into one trait. That is, would separate `Encoder` and `Decoder` traits make sense. As you say, though, if one isn't defined we could always return `Err` from the appropriate build function.

Reply:
Oh, I see. However, with that change we would now also support something like

```
codec = ["json", { type = "json", mode = "encode", field = "object", target_field = "serialized" }]
```

which would transform

```
"{ \"object\": { \"hello\": 123 } }"
```

into

```
{ "object": { "hello": 123 }, "serialized": "{ \"hello\": 123 }" }
```

This is quite contrived, but maybe it has some value?

> This is certainly more convenient for maintaining an inventory where there are both.

Other than that, this would be the main motivation 😄

Reply:
Changed that for now in 1f6a4c95c218dc783579dfd2e47a913d6b47af9b.

Reply:
Ok, I rolled it back now in 0cfc8278d85de87816d924d085be7b8c13bb2eea. I don't think anyone will ask for that feature 😄